accord-core/src/main/java/accord/topology/KeyRanges.java

```diff
@@ -38,11 +69,121 @@ public int size()
         KeyRange[] selection = new KeyRange[indexes.length];
         for (int i=0; i<indexes.length; i++)
             selection[i] = ranges[i];
         return new KeyRanges(selection);
     }
+
+    public boolean intersects(Keys keys)
+    {
+        for (int i=0; i<ranges.length; i++)
+            if (ranges[i].intersects(keys))
+                return true;
+        return false;
+    }
+
+    /**
+     * Subtracts the given set of key ranges from this
+     * @param that
+     * @return
+     */
+    public KeyRanges difference(KeyRanges that)
+    {
+        List<KeyRange> result = new ArrayList<>(this.size() + that.size());
+        int thatIdx = 0;
+
+        for (int thisIdx=0; thisIdx<this.size(); thisIdx++)
+        {
+            KeyRange thisRange = this.ranges[thisIdx];
+            while (thatIdx < that.size())
+            {
+                KeyRange thatRange = that.ranges[thatIdx];
+
+                int cmp = thisRange.compareIntersecting(thatRange);
+                if (cmp < 0)
+                {
+                    thatIdx++;
+                    continue;
+                }
+                if (cmp > 0) break;
+
+                int scmp = thisRange.start().compareTo(thatRange.start());
+                int ecmp = thisRange.end().compareTo(thatRange.end());
+
+                if (scmp < 0)
+                    result.add(thisRange.subRange(thisRange.start(), thatRange.start()));
+
+                if (ecmp <= 0)
+                {
+                    thisRange = null;
+                    break;
+                }
+                else
+                {
+                    thisRange = thisRange.subRange(thatRange.end(), thisRange.end());
+                    thatIdx++;
+                }
+            }
+            if (thisRange != null)
+                result.add(thisRange);
+        }
+        return new KeyRanges(result.toArray(KeyRange[]::new));
+    }
+
+    /**
+     * Adds a set of non-overlapping ranges
+     */
+    public KeyRanges add(KeyRanges that)
+    {
+        KeyRange[] combined = new KeyRange[this.ranges.length + that.ranges.length];
+        System.arraycopy(this.ranges, 0, combined, 0, this.ranges.length);
+        System.arraycopy(that.ranges, 0, combined, this.ranges.length, that.ranges.length);
+        Arrays.sort(combined, Comparator.comparing(KeyRange::start));
+
+        for (int i=1; i<combined.length; i++)
+            Preconditions.checkArgument(combined[i].compareIntersecting(combined[i -1]) != 0);
+
+        return new KeyRanges(combined);
+    }
+
+    public KeyRanges add(KeyRange range)
+    {
+        return add(new KeyRanges(new KeyRange[]{range}));
+    }
+
+    public KeyRanges mergeTouching()
+    {
+        if (ranges.length == 0)
+            return this;
+        List<KeyRange> result = new ArrayList<>(ranges.length);
+        KeyRange current = ranges[0];
+        for (int i=1; i<ranges.length; i++)
+        {
+            if (current.end().equals(ranges[i].start()))
```

Comment:
`end()` may be inclusive or exclusive. I think this is a problem in a few places.

Reply:
Yeah there are definitely a few assumptions here. Maybe merging should be implemented by the range itself